### PR TITLE
Add conditions to HTTPProxy and TLSCertificateDelegation

### PR DESCRIPTION
**Pull request by youngnick**
Updates #2495.

This adds a new `DetailedCondition` type to HTTPProxy, that allows for sub-conditions.

This allows Contour to have a `Valid` condition, that has further information about
how that conditions `status` has been reached, or in other words, to expose more
than one error or warning about a HTTPProxy to the user.

Also adds the Conditions block to TLSCertificateDelegation.

Signed-off-by: Nick Young <ynick@vmware.com>

### CODE REVIEW

File: apis/projectcontour/v1/kubeconditions.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+//     // Represents the observations of a foo's current state.
+//     // Known .status.conditions.type are: "Available", "Progressing", and "Degraded"
+//     // +patchMergeKey=type
+//     // +patchStrategy=merge
+//     // +listType=map
+//     // +listMapKey=type
+//     Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+//
+//     // other fields
+// }
```
Comment (jpeach) [inline apis/projectcontour/v1/kubeconditions.go:53]: The documentation generator really mangles this. It's adducing markdown, so you can probably tweak it just a little to come out legibly.

File: apis/projectcontour/v1/httpproxy.go [hunk 2]
```diff
@@ -467,6 +467,21 @@ type HTTPProxyStatus struct {
...
+	// +optional
+	// Conditions contains information about the current status of the HTTPProxy,
+	// in an upstream-friendly container.
+	// Contour will update a single condition, `Valid`, that is in normal-true polarity.
+	// That is, when `currentStatus` is `valid`, the `Valid` condition will be `status: true`,
+	// and vice versa.
+	// Contour will leave untouched any other Conditions set in this block,
+	// in case some other controller wants to add a Condition.
+	// If you are another controller owner and wish to add a condition, you *should*
+	// namespace your condition with a label, like `controller.domain.com\ConditionName`.
```
Comment (jpeach) [inline apis/projectcontour/v1/httpproxy.go:479]: Backslash seems like an unusual choice?

File: apis/projectcontour/v1/detailedconditions.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+//  The details of the errors will be present under the `errors` field.
+// There should never be subconditions under `errors` when `status` is `true`.
+type DetailedCondition struct {
+	// TODO(youngnick): Replace this with metav1.Condition once we have moved to a client-go
+	// version that includes it. Also includes deleting kubeconditions.go.
+	Condition `json:",inline"`
+	// Errors contains a slice of relevant error subconditions for this object.
+	// Subconditions are expected to appear when relevant (when there is a error), and disappear when not relevant.
+	// An empty slice here indicates no errors.
+	Errors []SubCondition `json:"errors,omitempty"`
```
Comment (jpeach) [inline apis/projectcontour/v1/detailedconditions.go:80]: Use `+optional` for both `Errors` and `Warnings` fields?

File: apis/projectcontour/v1/httpproxy.go (hunk 2 above)
Comment (youngnick) [reply apis/projectcontour/v1/httpproxy.go:479]: Oops, typo, thanks.

File: apis/projectcontour/v1/detailedconditions.go (hunk 3 above)
Comment (youngnick) [reply apis/projectcontour/v1/detailedconditions.go:80]: Yes, I intended to, thanks.

File: apis/projectcontour/v1/detailedconditions.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+// The expected lifecycle for these errors is that they should only be present when the error or warning is,
+// and should be removed when they are not relevant.
+type SubCondition struct {
+	// type of condition in CamelCase or in foo.example.com/CamelCase.
+	// ---
+	// Many .condition.type values are consistent across resources like Available, but because arbitrary conditions can be
+	// useful (see .node.status.conditions), the ability to deconflict is important.
+	// The regex it matches is (dns1123SubdomainFmt/)?(qualifiedNameFmt)
+	// +required
+	// +kubebuilder:validation:Required
```
[the remark this answers is not in the document]
Comment (youngnick) [reply apis/projectcontour/v1/detailedconditions.go:33]: Really? I just copied this from the upstream struct.

[the remark this answers is not in the document]
Comment (jpeach) [reply apis/projectcontour/v1/detailedconditions.go:33]: it doesn't do ant hard 🤷

**Pull request merged**